app/services/activitypub/fetch_remote_status_service.rb

```diff
@@ -4,36 +4,33 @@ class ActivityPub::FetchRemoteStatusService < BaseService
   include JsonLdHelper
 
   # Should be called when uri has already been checked for locality
-  def call(uri, prefetched_json = nil)
-    @json = body_to_json(prefetched_json) || fetch_resource(uri)
+  def call(uri, id: true, prefetched_body: nil)
+    if prefetched_body.nil?
+      @json = fetch_resource(uri, id)
+    else
+      @json = body_to_json(prefetched_body)
+    end
 
     return unless supported_context?
 
-    activity = activity_json
-    actor_id = value_or_id(activity['actor'])
-
-    return unless expected_type?(activity) && trustworthy_attribution?(uri, actor_id)
+    return if actor_id.nil? || !trustworthy_attribution?(@json['id'], actor_id)
 
     actor = ActivityPub::TagManager.instance.uri_to_resource(actor_id, Account)
-    actor = ActivityPub::FetchRemoteAccountService.new.call(actor_id) if actor.nil?
+    actor = ActivityPub::FetchRemoteAccountService.new.call(actor_id, id: true) if actor.nil?
 
     return if actor.suspended?
 
-    ActivityPub::Activity.factory(activity, actor).perform
+    ActivityPub::Activity.factory(activity_json, actor).perform
   end
 
   private
 
   def activity_json
-    if %w(Note Article).include? @json['type']
-      {
-        'type'   => 'Create',
-        'actor'  => first_of_value(@json['attributedTo']),
-        'object' => @json,
-      }
-    else
-      @json
-    end
+    { 'type' => 'Create', 'actor' => actor_id, 'object' => @json }
```

Comment:
Why did you change this? What if the link leads to an Announce or Move? Then this would wrap it inside Create and pass on to handle like a status without type validation..?

Reply:
`id` should not lead to an Activity if it is conforming to ActivityPub… I'm reviewing the implementation to figure out when it could receive an Activity.

Reply:
When we receive post / share activity via OStatus, we fetch AP version and use it if available.

https://github.com/tootsuite/mastodon/blob/c3f9c74719463be942b570c0a2e662775d33631c/app/lib/ostatus/activity/creation.rb#L12-L15

Then, FetchRemoteStatusService may be called with URI of an Announce activity. I think it's only case.

Reply:
Thank you for your quick answer. I'll amend the patch accordingly.

Reply:
In commit bfa808e0143f16ac6ce5dd1ff8bdc81e15148dd9, `OStatus::Activity::Creation` no longer delegates to ActivityPub because the provided ActivityPub representations are not signed while OStatus representations are.
If we still want the delegation, we need to sign activities in activity method of statuses controller, and to check the signature.